### ISSUE TITLE
Button groups with (split) dropdown/up

### ISSUE THREAD

**Issue by heruan** [vaadin/vaadin-button#86]
Common UI variations of buttons are:

- button groups
- button dropdowns/dropups (optionally split)

For example in Bootstrap:

<img width="729" alt="screen shot 2018-05-08 at 18 51 26" src="https://user-images.githubusercontent.com/851745/39771116-572d7bf4-52f1-11e8-85ee-cb44ddc8766b.png">

As shared with someone from the team I understand these should be somewhere on some roadmap, so I'm opening this to gather related discussion here.

**Comment by jouni** [comment]
Quick mockups for Lumo:
![menu-button and split-button](https://user-images.githubusercontent.com/66382/39772885-0fef8d46-52ff-11e8-8d33-a9e9bf52d120.png)

**Comment by jouni** [comment]
IMO we should reuse something like context-menu and vaadin-list-box here, so that we get keyboard navigation and other functionality for free. But we should first figure out a way to make something like a “vaadin-menu” element, which doesn’t always contain selectable items.

**Comment by jouni** [comment]
Not exactly sure which of these two issues is the main one.

This one here seems to have two use cases combined (maybe even three), while the referenced issue is just about grouping buttons visually (and perhaps about toggling the buttons on/off).

**Comment by heruan** [comment]
There is also https://github.com/vaadin/vaadin-select/issues/92 which might be a duplicate of this. There you said there was a `vaadin-menu-button` component planned: is this still true, or the feature will be in `vaadin-button`?

This is quite a blocker migrating from FW8 `MenuBar`, which I guess was used by many not as a real menu bar but more a dropdown button 🙂

**Comment by jouni** [comment]
To be honest, I’m not completely sure how the pieces will fall into place. But I’m quite sure we want to offer an explicit component for this use case, something like `<vaadin-menu-button>`. I don’t think this will be part of the select component, as the use case is different (a form field vs a navigation element).

And we’ve heard so much feedback about `MenuBar`, that it’s one of the priorities in our “bridging the gap” project (to make it possible for everyone to migrate from 7/8 -> 10). And I think the ”menu button” is the building block that a menu bar consists of.

**Comment by heruan** [comment]
If [these mockups](https://github.com/vaadin/vaadin-button/issues/86#issuecomment-387481409) will become a new `<vaadin-menu-button>` component, that would be ideal so from there it could evolve even more where `<vaadin-button>` wouldn't.

**Comment by heruan** [comment]
@jouni Why not having a `<vaadin-menu-dropdown>` instead, to be used as a dropdown component not only for buttons but also for other elements (e.g. avatars)?

**Comment by jouni** [comment]
Yeah, I agree, we should have a general purpose dropdown component that is then used to build the more opinionated components. The newish `<incubator-popup>` component is something like that: https://vaadin.com/components/incubator-popup

**Comment by heruan** [comment]
The `incubator-popup` component seems to fit the purpose, but I'm not sure how a commercial product could be used to provide a core feature such as a dropdown menu: I suppose a commercial component prevents its reusability in other elements, e.g. like `vaadin-combo-box` uses `vaadin-text-field`.

I didn't inspect each codebase yet, but I'd expect `vaadin-select`, `vaadin-como-box`, `vaadin-context-menu` and `incubator-popup` to share the same overlay/dropdown component already; then it should be quite straighforward to use the same on buttons, isn't it?

**Comment by jouni** [comment]
Yes, I would expect that as well, the popup/dropdown to be a shared, reusable part in all of those components.

But in reality, all of them implement the same feature separately. All of them use `<vaadin-overlay>`, but all the positioning and sizing logic are not shared.

The incubator-popup was custom built for a specific purpose in a customer project by a service developer, which partly explains why there’s not yet another dropdown component.

We’ll see when we have the time to implement an official general-purpose dropdown component (non-commercial) and after that start looking at migrating all the existing components to use that.

**Comment by heruan** [comment]
Any update/milestone for this?

**Comment by jouni** [comment]
We are working on the design this week together with @tomivirkki. The current idea is that we focus on the specific “menu bar” use case first (e.g. `File, Edit, View, ...`). We expect that we will end up building a `<vaadin-menu-button>`, at least internally, which we can package as a separate component after that.

**Comment by heruan** [comment]
As discussed in https://github.com/vaadin/vaadin-menu-bar-flow/issues/34 the `vaadin-menu-bar` isn't the right component to achieve this. So, this is what I've tried so far:

1. find a way to style a composite/group button
2. find a way to stick the context-menu to one of the buttons

I've failed at both: I've tried to set a custom theme to let two button appear adjacent, but I couldn't spot the style to override (is it a margin?); also tried positioning the context menu as described in https://github.com/vaadin/vaadin-context-menu/issues/153#issuecomment-455483333 but it didn't work for me (at least in Flow with `executeJs`).

**Comment by m-emelchenkov** [comment]
@jouni I love your mockups. This component is a must-have for Vaadin interfaces. There are some 3rd-party components but they does not look consistent in Vaadin Flow UI. I will await when this button will be available for use.

**Comment by sebdavid** [comment]
Hi,
Is this dropdown button component still in the pipe?
It seems it has been kind of "low prioritized" as it is now just a "component idea".

**Comment by web-padawan** [comment]
@sebdavid we have recently unified our GitHub labels across repos:

- `enhancement` label is used for new features to existing components, 
- `component-idea` means a totally new component which we might work on.

This component is not in the short term roadmap at the moment.
